akka-actor/src/main/scala/akka/io/TcpConnection.scala

```diff
@@ -70,9 +71,9 @@ private[io] abstract class TcpConnection(val tcp: TcpExt, val channel: SocketCha
 
       val info = ConnectionInfo(registration, handler, keepOpenOnPeerClosed, useResumeWriting)
 
-      // if we have resumed reading from pullMode while waiting for Register then register OP_READ interest
-      if (pullMode && !readingSuspended) resumeReading(info)
-      doRead(info, None) // immediately try reading, pullMode is handled by readingSuspended
+      // if we are in push mode or already have resumed reading in pullMode while waiting for Register
+      // then register OP_READ interest
+      if (!pullMode || ( /*pullMode && */ !readingSuspended)) resumeReading(info)
```

Comment:
Previously, we were calling `doRead` here directly, however, `doRead` may already try to close the connection and call `context.become(unregistering)` which would then be lost if `become` a few lines below.

It seems the `doRead` was an optimization in the first place trying to instantly read something without checking the selector first. Now, we always go through the selector, which is more consistent anyway.